win32ss/gdi/ntgdi/print.c

```diff
@@ -124,22 +87,69 @@ NtGdiExtEscape(
    INT    OutSize,
    OPTIONAL LPSTR  UnsafeOutData)
 {
-   PDC      pDC;
    LPVOID   SafeInData = NULL;
    LPVOID   SafeOutData = NULL;
    NTSTATUS Status = STATUS_SUCCESS;
    INT      Result;
+   PPDEVOBJ  ppdev;
+   PSURFACE psurf;
 
-   if (hDC == 0)
+   if (hDC == NULL)
    {
-       hDC = UserGetWindowDC(NULL);
+      if (pDriver)
+      {
+         /* FIXME : Get the pdev from its name */
+         UNIMPLEMENTED;
+         return -1;
+      }
+
+      ppdev = EngpGetPDEV(NULL);
+      if (!ppdev)
+      {
+         EngSetLastError(ERROR_BAD_DEVICE);
+         return -1;
+      }
+
+      psurf = ppdev->pSurface;
```

Comment:
This will require "EngAcquireSemaphore(ppdev->hsemDevLock);", otherwise there is a race with mode switching.

Reply:
Right.

Reply:
Can you confirm that the change suits you ?